### PR TITLE
feat(EMS-2424): Check your answers - Alternative trading address

### PR DESCRIPTION
**Pull request by Zainzzkk**
## Introduction :pencil2:
Adds check your answers for alternative trading address

## Resolution :heavy_check_mark:
- Added check your answers row for alternative trading address
- Conditionally revealed it if TRADING_ADDRESS is true
- Added logic to redirect to ALTERNATIVE_TRADING_ADDRESS_CHECK/CHANGE if TRADING_NAME is selected as true
- Added tests

## Miscellaneous :heavy_plus_sign:
- Added helper for generateAddressObject (moving from within summary list file)
- Added e2e command for checking row does not exist

### CODE REVIEW

File: e2e-tests/insurance/cypress/e2e/journeys/your-business/change-your-answers/change-your-answers-change-trading-address.spec.js [hunk 1]
```diff
@@ -0,0 +1,179 @@
...
+  EXPORTER_BUSINESS: {
+    COMPANY_DETAILS_CHANGE,
+    CHECK_YOUR_ANSWERS,
+    ALTERNATIVE_TRADING_ADDRESS_CHANGE,
+  },
+} = INSURANCE_ROUTES;
+
+const baseUrl = Cypress.config('baseUrl');
+
+context(`Insurance - Your business - Change your answers - ${TRADING_ADDRESS} and ${FULL_ADDRESS} - As an exporter, I want to change my answers to the trading address/ alternative trading address sections`, () => {
```
Comment (ttbarnes) [inline e2e-tests/insurance/cypress/e2e/journeys/your-business/change-your-answers/change-your-answers-change-trading-address.spec.js:39]: `trading address/ alternative`

File: e2e-tests/insurance/cypress/e2e/journeys/your-business/check-your-answers/check-your-answers-summary-list/check-your-answers-summary-list.spec.js [hunk 2]
```diff
@@ -37,68 +40,106 @@ context('Insurance - Your business - Check your answers - Summary list - your bu
...
-  });
+    it(`should render a ${TRADING_ADDRESS} summary list row`, () => {
+      checkSummaryList[TRADING_ADDRESS]();
+    });
 
-  it(`should render a ${WEBSITE} summary list row`, () => {
-    checkSummaryList[WEBSITE]();
-  });
+    it(`should not render a ${FULL_ADDRESS} summary list row`, () => {
+      checkSummaryList[FULL_ADDRESS](false);
```
Comment (ttbarnes) [inline e2e-tests/insurance/cypress/e2e/journeys/your-business/check-your-answers/check-your-answers-summary-list/check-your-answers-summary-list.spec.js:78]: What's the difference between this "false" flag and the new `assertSummaryListRowDoesNotExist` command?

File: src/ui/server/controllers/insurance/business/company-details/index.test.ts [hunk 3]
```diff
@@ -187,23 +189,54 @@ describe('controllers/insurance/business/companies-details', () => {
         expect(mapAndSave.companyDetails).toHaveBeenCalledWith(payload, mockApplication);
       });
 
-      describe(`when req.body has ${TRADING_ADDRESS} with a value of '${FIELD_VALUES.YES}'`, () => {
-        it(`should redirect to ${ALTERNATIVE_TRADING_ADDRESS_ROOT}`, async () => {
+      describe(`when req.body has ${TRADING_ADDRESS} with a value of 'true'`, () => {
+        it(`should redirect to ${ALTERNATIVE_TRADING_ADDRESS_ROOT} but is not a check or check-and-change route`, async () => {
```
Comment (ttbarnes) [inline src/ui/server/controllers/insurance/business/company-details/index.test.ts:193]: Seems the "but is not a ..." should be in the description

File: src/ui/server/controllers/insurance/business/company-details/index.test.ts [hunk 4]
```diff
@@ -187,23 +189,54 @@ describe('controllers/insurance/business/companies-details', () => {
...
+            [TRADING_ADDRESS]: 'true',
           };
 
           await post(req, res);
 
           const expected = `${INSURANCE_ROOT}/${mockApplication.referenceNumber}${ALTERNATIVE_TRADING_ADDRESS_ROOT}`;
           expect(res.redirect).toHaveBeenCalledWith(expected);
         });
+
+        it(`should redirect to ${ALTERNATIVE_TRADING_ADDRESS_CHANGE} but is check route`, async () => {
```
Comment (ttbarnes) [inline src/ui/server/controllers/insurance/business/company-details/index.test.ts:205]: same as above

File: src/ui/server/controllers/insurance/business/company-details/index.ts [hunk 5]
```diff
@@ -139,18 +147,32 @@ const post = async (req: Request, res: Response) => {
...
+    // if is a change route and TRADING_ADDRESS is true, then redirect ALTERNATIVE_TRADING_ADDRESS_CHANGE
+    if (submittedValues[TRADING_ADDRESS] && isChangeRoute(req.originalUrl)) {
+      changeRoute = `${INSURANCE_ROOT}/${referenceNumber}${ALTERNATIVE_TRADING_ADDRESS_CHANGE}`;
+    }
+
+    // if is a check-and-change route and TRADING_ADDRESS is true, then redirect ALTERNATIVE_TRADING_ADDRESS_CHECK_AND_CHANGE
+    if (submittedValues[TRADING_ADDRESS] && isCheckAndChangeRoute(req.originalUrl)) {
+      checkAndChangeRoute = `${INSURANCE_ROOT}/${referenceNumber}${ALTERNATIVE_TRADING_ADDRESS_CHECK_AND_CHANGE}`;
+    }
+
```
Comment (ttbarnes) [inline src/ui/server/controllers/insurance/business/company-details/index.ts:169]: Can we update these comments to be multi line? easier to read/digest so its "if ... then", e.g

```js

/**
 * If is a change route and TRADING_ADDRESS is true,
 * redirect ALTERNATIVE_TRADING_ADDRESS_CHANGE
 */
```

File: src/ui/server/helpers/summary-lists/generate-address-object/index.ts [hunk 6]
```diff
@@ -0,0 +1,23 @@
...
+  if (rawAddress) {
+    // replace new lines with line breaks to display in summary list
+    address = replaceNewLineWithLineBreak(rawAddress);
+  }
+
+  return {
+    address,
+  };
+};
+
```
Comment (ttbarnes) [inline src/ui/server/helpers/summary-lists/generate-address-object/index.ts:22]: Could reduce the lines here

```js
const generateAddressObject = (rawAddress?: string) =>
  rawAddress ? replaceNewLineWithLineBreak(rawAddress) : { DEFAULT.EMPTY };
```

File: e2e-tests/insurance/cypress/e2e/journeys/your-business/check-your-answers/check-your-answers-summary-list/check-your-answers-summary-list.spec.js (hunk 2 above)
Comment (Zainzzkk) [reply e2e-tests/insurance/cypress/e2e/journeys/your-business/check-your-answers/check-your-answers-summary-list/check-your-answers-summary-list.spec.js:78]: wanted to keep this test file constant.  So when you call checkSummaryList, it is able to handle this conditional row

**Review by ttbarnes — APPROVED** [review]

**Review by ttbarnes — APPROVED** [review]

**Pull request merged**